### PR TITLE
feat: update Siderail component and dependencies

### PR DESCRIPTION
**Pull request by karrui**
No more mobile view!!!!! 

### TL;DR
Refactored the `Siderail` component to improve code maintainability and usability. Updated the `type-fest` dependency from `4.21.0` to `4.23.0`.

### What changed?
- Refactored the `Siderail` component into multiple subcomponents (`SiderailLabel`, `SiderailList`, `SiderailParentLabel`) for better readability and maintainability.
- Introduced context for `Siderail` using `SiderailContext` for better state management.
- Updated the package version of `type-fest` from `4.21.0` to `4.23.0`.

### How to test?
- Ensure the Siderail component is rendering and functioning as expected in different layouts and interactions.
- Verify that the context is working correctly and shared states are consistent across different Siderail subcomponents.

### Why make this change?
This change aims to improve the maintainability of the Siderail component by breaking it down into smaller, reusable subcomponents, and to leverage context for better state management.

### CODE REVIEW

**Comment by karrui** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/398?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#398** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/398?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#385** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/385?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#384** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/384?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#370** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/370?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#359** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/359?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @karrui and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+const createSiderailLabelStyles = tv({
+  slots: {
+    container:
+      "group flex gap-1.5 px-3 py-4 transition hover:text-brand-interaction",
+    icon: "h-6 w-6 opacity-0 transition group-hover:translate-x-1 group-hover:opacity-100",
+  },
+  variants: {
+    isCurrent: {
+      true: {
+        container: "text-brand-interaction",
```
Comment (dcshzj) [inline packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:16]: nit: design issue but the contrast between the `text-brand-interaction` and the base text colour is very negligible, which make it difficult to tell exactly which page is the one that is the current page on the siderail

File: packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  return (
+    <LinkComponent
+      href={url}
+      className={compoundStyles.container({ className })}
+    >
+      {title}
+
+      {showIconOnHover && (
+        <BiRightArrowAlt aria-hidden className={compoundStyles.icon()} />
+      )}
```
Comment (dcshzj) [inline packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:54]: issue: the arrow's positioning on the user's browser doesn't seem to be consistent. One is beside the text and the other is at the end of the siderail. I think we should just put the arrow right at the end?

https://github.com/user-attachments/assets/03516088-68a4-4d98-8a51-e97ab80167d4

File: packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        {title}
+      </p>
+    )
+  }
+
+  return (
+    <LinkComponent
+      href={url}
+      className={compoundStyles.container({ className })}
+    >
```
Comment (dcshzj) [inline packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:49]: nit: related to the issue on arrow positioning, I think we should capture the user's arrow on the entire width of the siderail, not just for the text?

Screenshot 2024-07-31 at 17 34 43

as compared to 
Screenshot 2024-07-31 at 17 34 52

File: packages/components/src/templates/next/components/internal/Siderail/SiderailParentLabel.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+export const SiderailParentLabel = ({
+  parentTitle,
+  parentUrl,
+}: SiderailLabelProps) => {
+  const { LinkComponent } = useSiderailContext()
+
+  return (
+    <LinkComponent href={parentUrl} className={compoundStyles.container()}>
+      {parentTitle}
+      <BiRightArrowAlt aria-hidden className={compoundStyles.icon()} />
```
Comment (dcshzj) [inline packages/components/src/templates/next/components/internal/Siderail/SiderailParentLabel.tsx:27]: issue: same issue here regarding the arrow, should standardise with the other arrows in the siderail

File: packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx (hunk 1 above)
Comment (karrui) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:16]: ya told @sehyunidaaa liow she will give more think. Or just is the site theme's fault HAHA

File: packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx (hunk 3 above)
Comment (karrui) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:48]: made the arrow inline muaha. but the touch target will still be like that la it's flex wrap

File: packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import { BiRightArrowAlt } from "react-icons/bi"
+import { tv } from "tailwind-variants"
+
+import type { Item } from "./types"
+import { useSiderailContext } from "./SiderailContext"
+
+const createSiderailLabelStyles = tv({
+  slots: {
+    container: "group px-3 py-4 transition hover:text-brand-interaction",
+    icon: "-mt-0.5 inline h-6 w-6 opacity-0 transition group-hover:translate-x-1 group-hover:opacity-100",
```
Comment (dcshzj) [inline packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:10]: issue: eh I think it got worse

image

Comment (karrui) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:10]: expected right? now the icon sits with the line and breaks properly. that one just nice breakpoint ma. @sehyunidaaa right

Comment (sehyunidaaa) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:10]: oh errr shouldn't they be in a row! so the arrow is where the minus is and the text should wrap

Comment (karrui) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:10]: Omg so was always justify-between???? Ok

Comment (sehyunidaaa) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx:10]: ya my bad i never prototype the multi line version

Comment (karrui) [reply packages/components/src/templates/next/components/internal/Siderail/SiderailLabel.tsx]: done

File: packages/components/src/templates/next/components/internal/Siderail/SiderailList.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+const createSiderailListStyles = tv({
+  slots: {
+    header:
+      "flex w-full items-start justify-between gap-2 text-base-content-strong transition",
+    label: "prose-headline-base-semibold px-3 py-4",
+    childHeader: "flex w-full items-center gap-1.5",
+    childLabel: "py-2 hover:text-brand-interaction",
+    container: "",
+    expandButton: "py-4",
+    expandIcon: "h-6 w-6 transition hover:text-brand-interaction",
```
Comment (dcshzj) [inline packages/components/src/templates/next/components/internal/Siderail/SiderailList.tsx:28]: issue: hahaha I'm sorry but it is not aligned

Screenshot 2024-08-01 at 14 52 27
Screenshot 2024-08-01 at 14 53 06
Screenshot 2024-08-01 at 14 54 53


But okay lah at least the arrows themselves are aligned, just the plus minus signs

Screenshot 2024-08-01 at 14 53 53

**Review by dcshzj — APPROVED** [review]

**Comment by karrui** [comment]
### Merge activity

* **Aug 1, 11:59 PM EDT**: @karrui started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/isomer/398).

**Pull request merged**